### PR TITLE
ENH: add support for nan-like null strings in string replace

### PR DESCRIPTION
**Pull request by ngoldbaum**
This fixes an issue similar to the one fixed by #26353.

In particular, right now `np.strings.replace` calls the count ufunc to get the number of replacements. This is necessary for fixed-width strings, but it turns out to make it impossible to support null strings in replace.

I went ahead and instead found the replacement counts inline in the ufunc loop. This lets me add support for nan-like null strings, which it turns out pandas needs.

I marked this one as a backport and issued it separately from the other PR because the ufuncs fixed by the other PR aren't going to be in numpy 2.0.

### CODE REVIEW

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 1]
```diff
@@ -1330,11 +1332,29 @@ string_replace_strided_loop(
...
+                    if (has_nan_na) {
+                        if (NpyString_pack_null(oallocator, ops) < 0) {
+                            npy_gil_error(PyExc_MemoryError,
+                                          "Failed to deallocate string in replace");
+                            goto fail;
+                        }
+                        goto next_step;
+                    }
+                    else {
+                    npy_gil_error(PyExc_ValueError,
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1353]: Indentation off

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 2]
```diff
@@ -1330,11 +1332,29 @@ string_replace_strided_loop(
...
+                                  "Only nan-like null values are not supported "
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1354]: Delete "Only"?

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 3]
```diff
@@ -1349,32 +1369,52 @@ string_replace_strided_loop(
...
-        else {
-            // replace i2 with i3
-            max_size = i1s.size * (i3s.size/i2s.size + 1);
-        }
-        char *new_buf = (char *)PyMem_RawCalloc(max_size, 1);
-        Buffer<ENCODING::UTF8> buf1((char *)i1s.buf, i1s.size);
-        Buffer<ENCODING::UTF8> buf2((char *)i2s.buf, i2s.size);
-        Buffer<ENCODING::UTF8> buf3((char *)i3s.buf, i3s.size);
-        Buffer<ENCODING::UTF8> outbuf(new_buf, max_size);
+        {
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1372]: Why the new indentation? It already is in the loop.

(And it makes reviewing harder...)

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 4]
```diff
@@ -1349,32 +1369,52 @@ string_replace_strided_loop(
...
 
-        if (NpyString_pack(oallocator, ops, new_buf, new_buf_size) < 0) {
-            npy_gil_error(PyExc_MemoryError, "Failed to pack string in replace");
-            goto fail;
-        }
+            npy_int64 start = 0;
+            npy_int64 end = NPY_MAX_INT64;
 
-        PyMem_RawFree(new_buf);
+            npy_int64 found_count = string_count<ENCODING::UTF8>(buf1, buf2, start, end);
```
Comment (mhvk) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1384]: I'd just hard code `buf1, buf2, 0, NPY_MAX_INT64` - that seems clearer than defining variables that are only used here.

**Review by mhvk — COMMENTED** [review]
Looks good modulo some nitpicks...

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 3 above)
Comment (ngoldbaum) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:1372]: It's because of the new use of `goto next_step`, I need to defined a new lexical scope or define a bunch of variables at the top of the for loop that are only used at the bottom of it.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 2 above)
Comment (ngoldbaum) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp]: Thanks, fixed the double-negative and tweaked the wording. Hopefully the version I just pushed reads better.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 3 above)
Comment (mhvk) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:1372]: I'd probably have gone for top of the for-loop myself, but no big deal...

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 5]
```diff
@@ -1349,32 +1369,50 @@ string_replace_strided_loop(
...
+            // conservatively overallocate
+            // TODO check overflow
+            size_t max_size;
+            if (i2s.size == 0) {
+                // interleaving
+                max_size = i1s.size + (i1s.size + 1)*(i3s.size);
+            }
+            else {
+                // replace i2 with i3
+                max_size = i1s.size * (i3s.size/i2s.size + 1);
```
Comment (seberg) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1400]: That didn't change, but now that you have `count` you should use it, I think.

Also, am I confused by the division.  It seems correct, but a bit overly complicated, since you can use `i1s.size + difference` giving:
```
change = i2.size >= i3.size ? 0 : i3.size - i2.size;
max_size = i1s.size + count * change;
```
I.e. we replace at most count items (it might be less, if we can find overlaps with `string_count`.  If overlaps are impossible in `string_count` then I guess the count might be exact).

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 6]
```diff
@@ -1349,32 +1369,50 @@ string_replace_strided_loop(
...
+                in_count = NPY_MAX_INT64;
+            }
 
-        if (NpyString_pack(oallocator, ops, new_buf, new_buf_size) < 0) {
-            npy_gil_error(PyExc_MemoryError, "Failed to pack string in replace");
-            goto fail;
-        }
+            npy_int64 found_count = string_count<ENCODING::UTF8>(
+                    buf1, buf2, 0, NPY_MAX_INT64);
+            if (found_count == -2) {
```
Comment (seberg) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1383]:
```suggestion
            if (found_count < 0) {
```
Yes, it returns -2 due to `fastsearch`, but let's clarify that it can't actually return `-1`

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 3 above)
Comment (seberg) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:1372]: While I don't hate the `while (N--)` loop in general, I do think a `goto` for loop control flow isn't nice and I much prefer a long `for` instead.
But this file has this pattern in a few places right now so it doesn't matter since the other places use this pattern also.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp [hunk 7]
```diff
@@ -1330,11 +1332,29 @@ string_replace_strided_loop(
...
+                            npy_gil_error(PyExc_MemoryError,
+                                          "Failed to deallocate string in replace");
+                            goto fail;
+                        }
+                        goto next_step;
+                    }
+                    else {
+                        npy_gil_error(PyExc_ValueError,
+                                      "Only NaN-like null strings can be used "
+                                      "as search strings for replace");
```
Comment (seberg) [inline numpy/_core/src/umath/stringdtype_ufuncs.cpp:1355]: (just a curious note for now)

I think default strings don't actually hit this, right?  The only subtlety (which I don't care about), is that the we don't mutate the default string stored on the dtype probably, but rather insert the same string every time.

**Review by seberg — COMMENTED** [review]
Looks good to me too, and if we are in a rush, we could put it in.

However, what we are missing are tests for the error paths, I think even the now fixed nan-like null path is untested?

Unless I mind-slipped, I also think the size calculation is odd and should try to use `count` now?

File: numpy/_core/tests/test_stringdtype.py [hunk 8]
```diff
@@ -1218,6 +1218,7 @@ def test_unary(string_array, unicode_array, function_name):
     "strip",
     "lstrip",
     "rstrip",
+    "replace"
```
Comment (ngoldbaum) [inline numpy/_core/tests/test_stringdtype.py:1221]: @seberg this change makes sure the error paths are tested.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 7 above)
Comment (ngoldbaum) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:1355]: Ah good point; this error message isn't quite right, using a string as a missing string is also supported. Will update the error to match this.

Not sure what you're getting at about mutating strings, but that's why they're static strings that store the string data in a `const` buffer. Anyone mutating it is going out of their way to do so.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 5 above)
Comment (ngoldbaum) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp:1400]: Thanks. I agree this logic here is poorly motivated and using the `counts` directly makes more sense.

File: numpy/_core/src/umath/stringdtype_ufuncs.cpp (hunk 7 above)
Comment (seberg) [reply numpy/_core/src/umath/stringdtype_ufuncs.cpp]: I was thinking of:
```
dt1 = StringDType(na_value="hello")

replace(arr(..., dtype=dt1), "spam", "parrot")
```
doesn't give a `StringDtype(na_value="parrot")`, I think so "bloats" memory.

I don't mind that enough to worry (at least fo rnow, I think this is a niche feature)

**Comment by seberg** [comment]
Thanks for following up on the count also!
